### ISSUE TITLE
Static class causes IPython/pdb to freeze

### ISSUE THREAD

**Issue by spyder-issues-migrator** [spyder-ide/spyder#1931]
_From [gilboal...@gmail.com](https://code.google.com/u/108754594444438407981/) on 2014-08-10T16:59:13Z_

Spyder Version:  2.3.0beta2
Python Version:  3.3.5
Qt Version    :  4.8.4, PyQt4 (API v2) 4.9.6 on Windows
pyflakes >=0.6.0:  None (OK)
pep8 >=0.6      :  None (NOK)
IPython >=0.13  :  2.0.0 (OK)
rope >=0.9.2    :  0.9.4 (OK)
sphinx >=0.6.6  :  1.2.2 (OK)
matplotlib >=1.0:  1.3.1 (OK)
sympy >=0.7.0   :  0.7.5 (OK)
pylint >=0.25   :  1.1.0 (OK)

Background:
\-----------

I'm making my first steps in Python and have come to a point where I need a logging module. The reasons I didn't opt for a rotating file handler are:

a. I wanted a new folder to be created each time the code is run, hosting the new log files.

b. Using conventional filenames (myName_X.log, and not the default myName.log.X).

c. Limit log files by number of lines, and not file size (as done by the rotating file handler). 

What steps will reproduce the problem?

 --------------------------------------

1. Use the attached files. Run main_example.py either in run mode or in debug mode.

2. A new directory ("log_<time stamp>") is created under the current directory you are running in. A log file is created under that new directory.

3. Run main_example.py again. IPython (or pdb, in case you run in debug mode) hangs. The only solution is to close the console and open a new one. In addition, a second new directory ("log_<new time stamp>") is created under the current directory you are running in. A log file is created under that new directory. Try DELETING that file. On my machine, it is locked by python, and can be released only by closing Spyder, or releasing the file handle with an external tool (e.g. process explorer). 

What is the expected output? What do you see instead?

 -----------------------------------------------------

I expect IPython/pdb to continue running after the second time main_example.py is run. If there is a problem with the code, I'd expect to get some sort of message (I'm new to python, so it is likely to be the case).
Instead, the console freezes, and I must open a new one to continue working. 

Please provide any additional information below

. ------------------------------------------------

I'm running Windows 7 64bit. On Linux the problem doesn't happen. I tried using the logging module directly (on Win7), and it worked without any problems.

**Attachment:** [main_example.py myLogging.py](http://code.google.com/p/spyderlib/issues/detail?id=1931)

_Original issue: http://code.google.com/p/spyderlib/issues/detail?id=1931_

**Comment by spyder-issues-migrator** [comment]
_From [ccordoba12](https://code.google.com/u/ccordoba12/) on 2014-08-23T14:00:58Z_

**Labels:** -Cat-Console Cat-Debugger

**Comment by spyder-issues-migrator** [comment]
_From [ccordoba12](https://code.google.com/u/ccordoba12/) on 2014-08-24T23:09:09Z_

I don't understand too well what you're doing but:

a. This doesn't seem a Spyder bug but more a programming question for StackOverflow.

b. It's not possible to delete a file on Windows while it's in use. That's a Windows limitation and it doesn't have to do with Python, IPython or Spyder.

**Comment by spyder-issues-migrator** [comment]
_From [gilboal...@gmail.com](https://code.google.com/u/108754594444438407981/) on 2014-08-25T01:47:24Z_

I found the problem. It is a bug in the code - when an object of the static class was generated in the main function:

myLoggerInstance = myLogging.MyLogger()

the init method of the class added an additional handle to the logger:

MyLogger._logger.addHandler(MyLogger._hConsole)


This caused the handlers list to grow with each execution of main, and when I deleted the second element of the handler's list, it was no longer the file handler, but a console handler. 

MyLogger._logger.handlers[1].stream.close()

At this point I also realized that mutating a list using explicit indexing is a bad idea, and instead deleted the hFile handle itself.


@-ccordoba12, 

quote: "​a. This doesn't seem a Spyder bug but more a programming question for StackOverflow."

- I actually did publish this question on StackOverflow. I still think it is a bug in Spyder, because the behavior of the IDE should not be affected by code execution (I'd expect some exception to be raised and the "scope" returning back to IPython). Maybe when a fully integrated debugger will be incorporated into the IDE such a problem would easily be tracked down.


quote: "​b. It's not possible to delete a file on Windows while it's in use. "

- I expected python to release the handle to the file. With the bug fixed, the handle is released as expected.
